docs/src/references/c8y-configuration-management.md

```diff
@@ -201,71 +201,111 @@ To provide/consume configuration files to/from external devices, the HTTP filetr
+
+    * TO-BE-DEFINED:
+    
+      * Decide which error paths to be considered.
+      * Decide if any timeout to be considerd when waiting for upload notification from external device.
+
+  * C8Y config plugin: sends the obtained file to C8Y
+
+  * C8Y config plugin: removed the file form local filesystem
+  
 
 **Cloud-to-Device Behaviour:**
   * at some point a config sent from cloud for type `bar.conf` for `child1` arrives at C8Y config plugin<br/>
     Format of C8Y SmartREST message for config send operation: `524,<childid>,<URL>,<config type>`. See [C8Y SmartREST doc](https://cumulocity.com/guides/reference/smartrest-two/#download-configuration-file-with-type-524)<br/>
     Example: `524,child1,http://www.my.url,bar.conf`
-  * C8Y config plugin: stores the files received from C8Y to `/http-root/config/desired/child1/bar.conf`
-  * C8Y config plugin: notifies the external device about the new configuration via MQTT (reference to section Notifications above)
-  * external device child1: recognizes the MQTT notification and downloads the new config with an HTTP GET request
+  * C8Y config plugin: downloads the file based on the URL received from C8Y, and stores it some plugin specific local filesystem path (e.g. `/tmp/c8y-cfg-plugin/<childid>_<cfg-type>`).
+
+  * C8Y config plugin: uses the filetransfers local API to expose the file to the filetransfer feature. The filetransfer feature responses a _temporary URL_, that is later used to uniquely access that exposed file.
+
+    Example for a _temporary URL_: `/tedge/tmpfiles/<random string>`
+
+    * TODO: Investigate and decide about filetransfer's local API in scheduled prototype (https://github.com/thin-edge/thin-edge.io/issues/1307), according to topics below:
+ 
+      * local API must avoid that there are two copies of the same file at any point in time
+      * local API must avoid that the file content is transfered/copied from one to another place, when the plugin _exposes_ the file. 
+        Instead just the directory link may change (as it is the case for bash `mv` command).
+      * file is removed from plugin's source folder, when the file was successfully exposed.
+      
+      Reasons for requirements above: Saving disc-space and CPU resources for large files.
+
+  * C8Y config plugin: notifies the external device `child1` via MQTT to download the new configuration from the thin-edge device with the _temporary URL_.
+
+    Topic: `tedge/configuration/req/send/{config type}/{childid}`<br/>
+    Example: `tedge/configuration/req/send/bar.conf/child1`<br/>  
+    Payload: ` <temporary URL> `<br/>
+    TODO: Investigate and decide about topic structure and payload in scheduled prototype (https://github.com/thin-edge/thin-edge.io/issues/1307) 
+
+  * external device child1: Downloads configuration file from the thin-edge device with the HTTP filetransfer feature and the _temporary URL_, and filetransfer removes the file from it's local filesystem folder.
```

Comment:
Removing the file as soon as the download is complete sounds efficient in terms of storage, but I'm wondering if it'd lead to premature removal of this file after a partial download from the child device? Is it possible to detect if the file is "fully downloaded" by a client from the HTTP server side to achieve this? if not, wouldn't it be safer to remove the file after receiving the notification from the child device once it applies this config?

Reply:
I am not sure which kind of HTTP server/library we will use. However when the HTTP server has closed the file it has put the whole content into the TCP data stream. At that point the file can be removed. Lets see what information the HTTP library we will use give us about the transfer. And I expect whatever libary we will use it will give us at least an success or error signal as final result of the transfer - that would be then the signal to remove the file.

However I do not want to insist on that approach here. If it turns out that implementation-wise it is easier to get the indication via MQTT, we can do that. The important thing for me is, that the filetransfer takes care of the lifetime of that file, i.E. filetransfer is the one that removes the file.

Reply:
Decision was taken: Filetransfer will *not* remove the file, but the plugin is responsible for that. One reaon: A big firmware file may be re-used for more than one child-device; so more than one child-device shall be able to download it.